bentoml/_internal/resource.py

```diff
@@ -0,0 +1,279 @@
+                    quota = None
+        except FileNotFoundError as err:
+            logger.warning(f"Caught exception while calculating CPU quota: {err}")
+    if quota is not None and quota < 0:
+        quota = None
+    elif quota == 0:
+        quota = 1
+
+    if os.path.exists(shares_file):
+        try:
+            shares = _read_cgroup_file(shares_file) / float(1024)
+        except FileNotFoundError as err:
+            logger.warning(f"Caught exception while getting CPU shares: {err}")
+
+    os_cpu_count = float(os.cpu_count() or 1.0)
+
+    limit_count = math.inf
+
+    if quota and shares:
+        limit_count = min(quota, shares)
+    elif quota:
+        limit_count = quota
+    elif shares:
+        limit_count = shares
+
+    return float(min(limit_count, os_cpu_count))
+
+
+@functools.lru_cache(maxsize=1)
+def query_os_cpu_count() -> int:
+    cpu_count = os.cpu_count()
+    if cpu_count is not None:
+        return cpu_count
+    logger.warning("os.cpu_count() is NoneType")
+    return 1
+
+
+# class MemResource(Resource, resource_id="mem"):
+#     @classmethod
+#     def from_spec(cls, spec: t.Any):
+#         assert isinstance(mem, (int, str)), "mem must be int or str"
+#
+#         if isinstance(mem, int):
+#             return mem
+#
+#         unit_match = re.match("([0-9]+)([A-Za-z]{1,2})", mem)
+#         mem_multipliers = {
+#             "k": 1000,
+#             "M": 1000**2,
+#             "G": 1000**3,
+#             "T": 1000**4,
+#             "P": 1000**5,
+#             "E": 1000**6,
+#             "Ki": 1024,
+#             "Mi": 1024**2,
+#             "Gi": 1024**3,
+#             "Ti": 1024**4,
+#             "Pi": 1024**5,
+#             "Ei": 1024**6,
+#         }
+#         if unit_match:
+#             base = int(unit_match[1])
+#             unit = unit_match[2]
+#             if unit in mem_multipliers:
+#                 return base * mem_multipliers[unit]
+#         raise ValueError(f"Invalid MEM resource limit '{mem}'")
+
+
+class NvidiaGpuResource(Resource[int], resource_id="nvidia.com/gpu"):
+    @classmethod
+    def from_spec(cls, spec: t.Any) -> int:
+        if not isinstance(spec, (int, str)):
+            raise TypeError("nvidia.com/gpu must be int or str")
+
+        if isinstance(spec, int):
+            return spec
+        return int(spec)
+
+    @classmethod
+    def from_system(cls) -> int:
+        """
+        query nvidia gpu count, available on Windows and Linux
+        """
+        import pynvml  # type: ignore
+
+        try:
+            pynvml.nvmlInit()
+            device_count = pynvml.nvmlDeviceGetCount()
+            return device_count
+        except (pynvml.nvml.NVMLError, OSError):
+            return 0
+        finally:
+            try:
+                pynvml.nvmlShutdown()
+            except Exception:  # pylint: disable=broad-except
+                pass
+
+
+def get_gpu_memory(dev: int) -> t.Tuple[float, float]:
```

Comment:
Is this being used?

Reply:
No, I don't believe so.

Reply:
This is useful for our GPU strategy. We should just put it here for now.